perun/collect/trace/strategy.py

```diff
@@ -28,213 +28,287 @@ class _Status(IntEnum):
+                             return the expected ending suffix or collection of them
+    :return list: the list of combined suffix pairs
     """
-    # Custom has higher priority than extracted
-    seen = set()
-    for probe in custom:
-        seen.add(probe['name'])
-    for probe in extracted:
-        if probe['name'] not in seen:
-            custom.append(probe)
-    return custom
+    # Check if the probe has matching starting and ending suffixes
+    pairs = []
+    for suffix in list(b_suffixes):
+        pair_suffixes = pairing(suffix[0].lower())
+        # Can't use simple 'in' - we need case insensitive comparison and case sensitive result
+        for e_suffix in list(e_suffixes):
+            if e_suffix[0].lower() in pair_suffixes:
+                b_suffixes.remove(suffix)
+                e_suffixes.remove(e_suffix)
+                pairs.append((suffix[1] + suffix[0], e_suffix[1] + e_suffix[0]))
+                break
+    return pairs
+
+
+# TODO: Allow multiple pairs (needs advanced processing of sampling across all the pairs)
+def _add_paired_rule(probe_start, probe_end, start_sample, end_sample, result, add_separate=True):
+    """Add paired rule to the resulting probe dictionary.
+
+    :param str probe_start: the full name of starting probe
+    :param str probe_end: the full name of paired ending probe
+    :param int start_sample: the sampling value for starting probe
+    :param int end_sample: the sampling value for ending probe
+    :param dict result: the dictionary of static probes
+    :param bool add_separate: if one of the paired probe is already in the result, add the
+                              remaining probe as simple non-paired probe
+    """
+    # Insert the created pair if both probes are unique (multiple pairing problem)
+    if probe_start not in result and probe_end not in result:
+        sampling = min(start_sample, end_sample)
+        result[probe_start] = {'name': probe_start, 'sample': sampling,
+                               'pair': (RecordType.StaticBegin, probe_end)}
+        result[probe_end] = {'name': probe_end, 'sample': sampling,
+                             'pair': (RecordType.StaticEnd, probe_start)}
+    elif add_separate:
+        # The probes are not unique, do the single insertion by setdefault
+        result.setdefault(probe_start, {'name': probe_start, 'sample': start_sample})
+        result.setdefault(probe_end, {'name': probe_end, 'sample': end_sample})
+
+
+def _add_suffix_rules(name, suffixes, sample, result):
+    """Add static probe rules created from name and suffixes as non-paired rules
+
+    :param str name: the base name of the probe
+    :param list suffixes: list of all suffixes to add
+    :param int sample: the sampling value for the rules
+    :param dict result: the static probe dictionary
+    """
+    if not suffixes:
+        result.setdefault(name, {'name': name, 'sample': sample})
+    for suffix in suffixes:
```

Comment:
Suggestion: Again this could be extracted directly in the for loop (i.e. for (suffix_0, suffix_1) in suffixes)